graylog2-web-interface/src/views/components/contexts/WidgetFocusProvider.tsx

```diff
@@ -15,67 +15,123 @@
  * <http://www.mongodb.com/licensing/server-side-public-license>.
  */
 import * as React from 'react';
-import { useState, useEffect, useCallback } from 'react';
+import { isEqual } from 'lodash';
+import { useState, useEffect, useCallback, useMemo } from 'react';
 import { useLocation, useHistory } from 'react-router-dom';
 import PropTypes from 'prop-types';
 import URI from 'urijs';
 
 import { useStore } from 'stores/connect';
 import useQuery from 'routing/useQuery';
 import { WidgetStore } from 'views/stores/WidgetStore';
-import WidgetFocusContext from 'views/components/contexts/WidgetFocusContext';
+import WidgetFocusContext, { FocusRequest } from 'views/components/contexts/WidgetFocusContext';
 
-const _syncWithQuery = (query: string, focusedWidget: string) => {
-  const baseUri = new URI(query)
-    .removeSearch('focused');
+type QueryParamsUpdate = FocusRequest & { id: FocusRequest['id'] | undefined };
 
-  if (focusedWidget) {
-    return baseUri.setSearch('focused', focusedWidget).toString();
+const _updateQueryParams = ({
+  newQueryParams,
+  query,
+} : {
+  newQueryParams: QueryParamsUpdate,
+  query: string,
+}) => {
+  let baseUri = new URI(query)
+    .removeSearch('focusing')
+    .removeSearch('editing')
+    .removeSearch('focusedId');
+
+  if (newQueryParams?.id && (newQueryParams.focusing || newQueryParams.editing)) {
+    baseUri = baseUri.setSearch('focusedId', newQueryParams.id);
+  }
+
+  if (newQueryParams.focusing) {
+    baseUri = baseUri.setSearch('focusing', true);
+  }
+
+  if (newQueryParams.editing) {
+    baseUri = baseUri.setSearch('editing', true);
   }
 
   return baseUri.toString();
 };
 
-const useFocusWidgetIdFromParam = (focusedWidget, setFocusedWidget, widgets) => {
-  const { focused: paramFocusedWidget } = useQuery();
-
+const useSyncStateWithQueryParams = ({ focusedWidget, focusUriParams, setFocusedWidget, widgets }) => {
   useEffect(() => {
-    if (focusedWidget !== paramFocusedWidget) {
-      if (paramFocusedWidget && !widgets.has(paramFocusedWidget)) {
+    const nextFocusedWidget = {
+      id: focusUriParams.id,
+      editing: focusUriParams.editing,
+      focusing: focusUriParams.focusing || focusUriParams.editing,
+    };
+
+    if (!isEqual(focusedWidget, nextFocusedWidget)) {
+      if (focusUriParams.id && !widgets.has(focusUriParams.id)) {
         return;
       }
 
-      setFocusedWidget(paramFocusedWidget);
+      setFocusedWidget(nextFocusedWidget);
+    }
+  }, [focusedWidget, setFocusedWidget, widgets, focusUriParams]);
+};
+
+const useCleanupQueryParams = ({ focusUriParams, widgets, query, history }) => {
+  useEffect(() => {
+    if ((focusUriParams?.id || focusUriParams?.editing || focusUriParams?.focusing) && !widgets.has(focusUriParams.id)) {
```

Comment:
I think the intention here is to clear parameters if a) an invalid widget id is focused or b) if `editing`/`focusing` are specified, but no `id`. If this is the case, we can probably make this a bit more explicit by stating this as:

```suggestion
    if ((focusUriParams?.id && !widgets.has(focusUriParams.id) 
        || (focusUriParams?.id === undefined)) {
```